Oracle PLSQL lexer (suitable for Oracle SQL as well)

lee-lindley opened the pull request:
The keywords, non-reserved keywords, types and built-in functions are from the Oracle 19c manuals.
The quoted literal parsing is much better than the existing SQL lexer, especially regarding the "q'" operator. I classify the quote delimiters as operators which is a departure from most others, but they are not part of the literal, so making them the same color as the quoted string is just wrong.

I borrowed liberally from the existing SQL, Perl and Ruby lexers and folowed the Lexer Development guide.
The demo, sample and spec files are similar to others in those directories.

My blog (http://lee-lindley.github.io) is built using this lexer.

tancnle commented on lib/rouge/lexers/plsql.rb:
```diff
@@ -0,0 +1,596 @@
+      end
+
+      state :dotnames do
+            # if we are followed by a dot and another name, we are an ordinairy name
```
**nitpick:** Typo?

```suggestion
            # if we are followed by a dot and another name, we are an ordinary name
```

tancnle commented on lib/rouge/lexers/plsql.rb:
```diff
@@ -0,0 +1,596 @@
+        TRANSACTIONAL TRIGGER UB1 UB4 UNION UNIQUE UNSIGNED UNTRUSTED UPDATE VALIST
+        VALUES VARIABLE VIEW VOID WHERE WHILE WITH
+        ))
+      end
+      def self.keywords_nresvd
```
**nitpick:** Could we add a line separator between these two functions for better readability?

tancnle replied:
**question:** Can we rename this method to `keywords_reserved` for consistency with the other lexers? Or does this name refer to something totally different? 🤔

tancnle commented on lib/rouge/lexers/plsql.rb:
```diff
@@ -0,0 +1,596 @@
+        XMLTOKENSET XMLTRANSFORM XMLTRANSFORMBLOB XMLTSET_DML_ENABLE XML_DIAG XML_DML_RWT_STMT
+        XPATHTABLE XS XS_SYS_CONTEXT X_DYN_PRUNE YEARS YES ZONEMAP
+        ))
+      end
+      def self.keywords_func
```
**nitpick:** Same as above. Could we add a blank line between function blocks?

tancnle commented on lib/rouge/lexers/plsql.rb:
```diff
@@ -0,0 +1,596 @@
+      end
+
+      def self.keywords_type
+        # sources:
```
**suggestion:** Is this comment needed?

tancnle commented on lib/rouge/lexers/plsql.rb:
```diff
@@ -0,0 +1,596 @@
+        rule %r/--.*/, Comment::Single
+        rule %r(/\*), Comment::Multiline, :multiline_comments
+        # Q' operator quoted string literal
+        rule %r/q'(.)/i  do |m|
```
```suggestion
        rule %r/q'(.)/i do |m|
```

tancnle commented on lib/rouge/lexers/plsql.rb:
```diff
@@ -0,0 +1,596 @@
+        rule %r(/\*), Comment::Multiline, :multiline_comments
+        # Q' operator quoted string literal
+        rule %r/q'(.)/i  do |m|
+            #open = Regexp.escape(m[1])
```
**question:** Redundant comment?

tancnle commented on lib/rouge/lexers/plsql.rb:
```diff
@@ -0,0 +1,596 @@
+        rule %r/(\$(?:IF|THEN|ELSE|ELSIF|ERROR|END|(?:\$\$?\w[\w\d]*)))(\s+)/im do
+            groups Comment::Preproc, Text
+        end
+        ### we do not use backticks in Oracle. I do not know the rules for other sql engines
+        ###rule %r/`/, Name::Variable, :backtick
+
```
**suggestion:** I think we could drop these comments. What do you think?

tancnle commented on lib/rouge/lexers/plsql.rb:
```diff
@@ -0,0 +1,596 @@
+        rule %r/(\w[\w\d\$]*)(\.(?=\w))?/ do |m|
+          if self.class.keywords_type.include? m[1].upcase
+            tok = Keyword::Type 
+            #Name::Builtin
```
**question:** Redundant comment?

tancnle commented on lib/rouge/lexers/plsql.rb:
```diff
@@ -0,0 +1,596 @@
+            #Name::Builtin
+          elsif self.class.keywords_func.include? m[1].upcase
+            tok = Name::Function
+          elsif self.class.keywords.include? m[1].upcase
+            tok = Keyword::Reserved
+          elsif self.class.keywords_nresvd.include? m[1].upcase
+            tok = Keyword
```
**suggestion:** Could we swap the token types?

```suggestion
          elsif self.class.keywords.include? m[1].upcase
            tok = Keyword
          elsif self.class.keywords_nresvd.include? m[1].upcase
            tok = Keyword::Reserved
```

Or I could have misread the intention here.

tancnle commented on lib/rouge/lexers/plsql.rb:
```diff
@@ -0,0 +1,596 @@
+      end
+
+
+      #state :backtick do
+      #  rule %r/\\./, Str::Escape
+      #  rule %r/``/, Str::Escape
+      #  rule %r/`/, Name::Variable, :pop!
+      #  rule %r/[^\\`]+/, Name::Variable
+      #end
```
**suggestion:** If backtick is not supported in PL/SQL, can we remove this block altogether instead of commenting?

lee-lindley replied:
Thanks for the detailed review Tan!
I'll address all of this as time permits. It shouldn't take long (though not today).
Unfortunately, if I dive back in I'll be tempted to tweak some more. I'll try to restrain myself.

tancnle reviewed (commented):
Thank you for your contribution and hard work @lee-lindley ❤️ The PR looks great to me overall 👍🏼 I have added some suggestions/questions for your consideration.

- Could we ensure 2 space indentation on the `lib/rouge/lexers/psql.rb` file? There seem to be some mixed 2 and 4 space indentations on the file.
- Could we add an entry on the `docs/Languages.md` file for this new lexer?

Please let me know what you think 🙏🏼 .

lee-lindley commented:
Tan,
I answered each one as I was editing and pressed the "resolved" button. Because I do not see them here, I'm thinking github dumped my comments when I did that instead of saving it along with the resolution. Lesson learned.

In any case I believe I addressed all suggestions appropriately, and only did small amount of functionality tweaking. I reviewed my test files visually and everything still looks good.

tancnle commented:
Thanks a lot for your prompt action @lee-lindley 🙇🏼 Looks great to me 🚀
